reference/src/representation/structs-and-tuples.md

```diff
@@ -0,0 +1,370 @@
+`std::marker::PhantomData`.
+
+[#37]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/37
+
+**Single-field structs ([#34]).** If you have a struct with single field
+(`struct Foo { x: T }`), should we guarantee that the memory layout of
+`Foo` is identical to the memory layout of `T` (note that ABI details
+around function calls may still draw a distinction, which is why
+`#[repr(transparent)]` is needed). What about zero-sized types like
+`PhantomData`?
+
+[#34]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/34
+
+**Homogeneous structs ([#36]).** If you have homogeneous structs, where all
+the `N` fields are of a single type `T`, can we guarantee a mapping to
+the memory layout of `[T; N]`? How do we map between the field names
+and the indices? What about zero-sized types?
+
+[#36]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/36
+
+**Deterministic layout ([#35]).** Can we say that layout is some deterministic
+function of a certain, fixed set of inputs? This would allow you to be
+sure that if you do not alter those inputs, your struct layout would
+not change, even if it meant that you can't predict precisely what it
+will be. For example, we might say that struct layout is a function of
+the struct's generic types and its substitutions, full stop -- this
+would imply that any two structs with the same definition are laid out
+the same. This might interfere with our ability to do profile-guided
+layout or to analyze how a struct is used and optimize based on
+that. Some would call that a feature.
+
+[#35]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/35
+
+### C-compatible layout ("repr C")
+
+For structs tagged `#[repr(C)]`, the compiler will apply a C-like
+layout scheme. See section 6.7.2.1 of the [C17 specification][C17] for
+a detailed write-up of what such rules entail (as well as the relevant
+specs for your platform). For most platforms, however, this means the
+following:
+
+[C17]: http://www.open-std.org/jtc1/sc22/wg14/www/abq/c17_updated_proposed_fdis.pdf
+
+- Field order is preserved.
+- The first field begins at offset 0.
+- Assuming the struct is not packed, each field's offset is aligned[^aligned] to
+  the ABI-mandated alignment for that field's type, possibly creating
+  unused padding bits.
+- The total size of the struct is rounded up to its overall alignment.  
+
+[^aligned]: Aligning an offset O to an alignment A means to round up the offset O until it is a multiple of the alignment A.
+
+One deviation from C comes about with "empty structs". In Rust, a
+struct that contains (transitively) no data members is considered to
+have size zero, which is not something that exists in C. This includes
+a struct like `#[repr(C)] struct Foo { }`. Further, when a
+`#[repr(C)]` struct has a field whose type has zero-size, that field
+may induce padding due to its alignment, but will not otherwise affect
+the offsets of subsequent fields (as it takes up zero space).
```

Comment:
As @rkruppe noted on Zulip, this seems like a problem. It means that "copy-pasting `struct` definitions and adding `repr(C)` everywhere" does not give you C compatibility, because your `Foo` would actually take space when put in a larger struct in `C`.

This seems like a bug, TBH. I am not sure if it is a bug that we can still fix. Might be worth having at least a warning.

Reply:
Yes — I was thinking the same in that conversation. That is, "bug and not clearly a bug we can fix", which does suggest that at least a lint is warranted.

Reply:
For the benefit of others, the [Zulip conversation was here](https://rust-lang.zulipchat.com/#narrow/stream/136281-wg-unsafe-code-guidelines/subject/how.20much.20determinism/near/136350308).

Reply:
That's not actually how C works - C does not allow empty structs. The godbolt link was from C++, which does allow empty structs, and in which empty structs have size one.

Reply:
Transcribing another thing @ubsan noted on Zulip: empty structs are accepted as an extension by some C compilers, but (at least) GCC and Clang make them have size zero, unlike C++. Example: https://godbolt.org/z/AS2gdC

Reply:
Oh wow. So this same data structure behaves different in C and C++.

Seems like the most sensible choice then is to keep size 0 in Rust (consistent with the rest of Rust, and with C in at least GCC and Clang), but still have a lint about the difference to C++?

Reply:
Allow-by-default at best - 0 size structures are weird in C++, and usually you'd use either EBO or `[[no_unique_address]]` with them.

Reply:
Isn't them being weird another argument for making this a warn-by-default lint?  I expect many people will not know this.  I am not a C++ expert, but I have programmed in C++ for many years and never heard about this; I don't think we can expect everybody doing C++ FFI to know about these issues.

Reply:
So to summarize: C does not allow empty structs, some C language extensions allow empty structs with `sizeof == 0`, C++ does allow empty structs but these have a `sizeof == 1` unless they are inherited from (in which case EBO guarantees that they won't increase the size of the struct) or they are fields that have the `[[no_unique_address]]` attribute. 

I think `#[repr(C)]` should warn on this when it makes a difference by default, that is, when the ZST would change the layout. We could have an opt-in warning that always warns on ZST being used in `#[repr(C)]` but I fear that would be extremely noisy for little win. 

About the situation with the C-language extension and C++ it appears that `#[repr(C)] != #[repr(C++)]`, so maybe we just need to add new `repr`s to deal with those.

Reply:
I had to google EBO: [Empty base optimization](https://en.cppreference.com/w/cpp/language/ebo).

Reply:
@RalfJung note that EBO is guaranteed by the C++>=11 standard for types with [standard layout](https://en.cppreference.com/w/cpp/named_req/StandardLayoutType), like empty structs: `struct T {};`, so it is a required layout optimization.